pmd-core/src/main/java/net/sourceforge/pmd/cache/AbstractAnalysisCache.java

```diff
@@ -79,6 +79,13 @@ public boolean isUpToDate(final TextDocument document) {
 
             if (result) {
                 LOG.debug("Incremental Analysis cache HIT");
```

Comment:
In `updatedResultsCache` are all the violations, that are persisteted in the end (by FileAnalysisCache#persist).
When we have a cache hit, where do we actually add the cached violations to the `updatedResultsCache` so that it is persisted again? Here we created an empty updatedResult. New violations are added via the FileAnalysisListener, but I don't see the code for cached violations.

I've now tested it a bit:
* first run of PMD with empty cache. Violations are reported and added to the cache
* second run of PMD with filled cache. Violations are reported (from cache) and the updated cache removes the violation.
* third run of PMD with updated cache. No violations are reported (from cache). 

I think, we need here something like this:

```
updatedResult.addViolations(analysisResult.getViolations());
```

I wonder, why we didn't catch this earlier in tests? Maybe we need a test at the CLI level - kind of a integration test, that calls PMD at least twice with cache?

Reply:
In PMD 6 at least violations are to be retrieved by the processor and reported back. That way all listeners (in particular report listeners) get to know about them. I will double check this is working as expected.

Reply:
It seems in PMD 7, we explicitly *not* call the cache listener:

https://github.com/pmd/pmd/blob/0547d907b0c4784ef497c8763946c567bfd76ff4/pmd-core/src/main/java/net/sourceforge/pmd/processor/PmdRunnable.java#L82-L84

Reply:
When I wrote that logic I think I assumed that if the cache listener got fed the violations it fetched from the cache, that would duplicate them. I assumed the cache wouldn't rebuild a cache entry from the listener, but just keep the existing cache entry around to persist it. Maybe that's not working properly

Reply:
yeah, that's the issue. I'll fix it, and add an integration test under `pmd-dist` asserting a report from cache is identical to one without it.